sqlair.go

```diff
@@ -453,6 +477,21 @@ func (tx *TX) Query(ctx context.Context, s *Statement, inputArgs ...any) *Query
 	}
 
 	run := func(innerCtx context.Context) (rows *sql.Rows, result sql.Result, err error) {
+		sqlstmt, ok := stmtCache.lookupStmt(tx.db, s, pq.SQL())
+		if ok {
+			// Register the prepared statement on the transaction. This function
+			// does not resend the prepare request to the database.
+			// The txstmt is closed by database/sql when the transaction is
+			// commited or rolled back.
+			txstmt := tx.sqltx.Stmt(sqlstmt)
+			if pq.HasOutputs() {
+				rows, err = txstmt.QueryContext(innerCtx, pq.Params()...)
+			} else {
+				result, err = txstmt.ExecContext(innerCtx, pq.Params()...)
+			}
+			return rows, result, err
+		}
```

Comment:
Under the original https://github.com/canonical/sqlair/pull/81 we were re-preparing each time because the statement was closed with the transaction.

In attempting to fix this under https://github.com/canonical/sqlair/pull/117, it was noted that preparing against the transaction's database required an extra connection to be acquired, which opened the possibility of deadlock for situations where [SetMaxOpenConns](https://pkg.go.dev/database/sql#DB.SetMaxOpenConns) had been called with a some number > 0.

Why don't we check the `DBStats` here and do this preparation if `MaxOpenConnections` is zero? Juju would then be able to benefit from the statement caching.

Reply:
We did consider that, but it feels very unsafe in regard to concurrency. For one, the SetMaxOpenConns could be set between us reading the DBstats and trying to open the connection. Secondly we do not control the `sql.db`. When the `sqlair.DB` is created we just use the `sql.db` passed to it. It is possible for another reference of that to be held separately and for someone to call `MaxOpenConns` on that without our knowledge.